Add C# Asynchronous Programming Async/Await Chapter

xantho09 opened the pull request:
Closes #194

xantho09 committed:
Add C# Async/Await Chapter

damithc reviewed (approved):
Structure seems to be fine. Did not spot any content problems when I did quick read. Let's wait one more day to see if there are any other reviews.

j-lum commented on contents/csharp/CSAsyncAwait.md:
```diff
@@ -0,0 +1,322 @@
+
+## What Is Asynchronous Programming?
+
+In **synchronous programming**, code is executed one step at a time, waiting for each step to be finished before proceeding to the next step. For example, if the code looks something like this:
```
Might want to clarify whether the steps are dependent on each other/need to be return their results in a specified order for this example.

Might also want to clarify that async can result in steps finishing in a different order.

j-lum commented on contents/csharp/CSAsyncAwait.md:
```diff
@@ -0,0 +1,322 @@
+5. Use the result from Task 1 to start Task 4.
+6. Wait for Tasks 2, 3 and 4 to finish.
+
+This is the main idea behind asynchronous programming: while waiting for tasks to complete in the background, perform other tasks instead of just waiting there and doing nothing. This can result in better utilization of resources at hand.
```
Well technically the current task is being executed. Is there a layman's way to explain execution contexts or just mention threading?

xantho09 replied:
Admittedly, the introduction was the hardest part to write. It was tricky to keep the explanation simple while making sure it gets the idea across.

I think mentioning threading is probably the simplest solution. Something along the lines of:
- You have 3 additional threads, A, B and C.
- Start Task 1 on Thread A.
- Start Task 2 on Thread B.
And so on.

xantho09 replied:
How does it look now?

I'm not too satisfied with the sentence that follows _"This is the main idea behind asynchronous programming"_; I feel like there's a better way to phrase it.

j-lum commented on contents/csharp/CSAsyncAwait.md:
```diff
@@ -0,0 +1,322 @@
+
+### Potential Pitfall: Deadlock
+
+It takes some time to understand the correct usage of `async` and `await`. If you're not careful, you could end up with a deadlock situation. Consider the following example:
```
I kind of got the idea after reading the article provided. Do you think having a similar diagram from the control flow section could help illustrate the idea here?

xantho09 replied:
That might be a good idea.

I think I should probably briefly mention (possibly in the `async` section)
- Why the completed task needs to wait for the UI thread specifically (and not just any thread in the thread pool)
- How the execution resumes _at the first opportunity_ the thread is available. (Since that availability never comes, the execution can't resume)

j-lum reviewed (commented):
Content-wise looks mostly OK to me. My background in async is mainly from scala so I can't really comment too much on the C# code but they look simple and sufficient.

xantho09 committed:
Add deadlock diagram and fix introduction

xantho09 committed:
Add reviewers' names

xantho09 commented:
@damithc It doesn't seem like @j-lum is responding to the comments he left, but I've addressed the points he made. Reviewers' names have been added. 

It should be ready for merging.

damithc commented:
@xantho09 Will merge soon. But first, can you remove my name from the reviewers list, to be consistent with other pages?

xantho09 committed:
Remove Prof Damith's name from reviewers

xantho09 commented:
Done.

Please and thank you.
